includes/sanitizers/class-amp-meta-sanitizer.php

```diff
@@ -0,0 +1,333 @@
+
+	/**
+	 * Tag.
+	 *
+	 * @var string HTML <meta> tag to identify and replace with AMP version.
+	 */
+	public static $tag = 'meta';
+
+	/**
+	 * Placeholder for default arguments, to be set in child classes.
+	 *
+	 * @var array
+	 */
+	protected $DEFAULT_ARGS = [ // phpcs:ignore WordPress.NamingConventions.ValidVariableName.PropertyNotSnakeCase
+		'use_document_element' => true, // We want to work on the header, so we need the entire document.
+	];
+
+	/**
+	 * Reference to the shared XPath object to query the DOM.
+	 *
+	 * @var DOMXPath
+	 */
+	protected $xpath;
+
+	/**
+	 * The document's <head> element.
+	 *
+	 * @var DOMElement
+	 */
+	protected $head;
+
+	/*
+	 * Tags array keys.
+	 */
+	const TAG_CHARSET        = 'charset';
+	const TAG_VIEWPORT       = 'viewport';
+	const TAG_AMP_SCRIPT_SRC = 'amp_script_src';
+	const TAG_OTHER          = 'other';
+
+	/**
+	 * Associative array of DOMElement arrays.
+	 *
+	 * Each key in the root level defines one group of meta tags to process.
+	 *
+	 * @var array $tags {
+	 *     An array of meta tag groupings.
+	 *
+	 *     @type DOMElement[] $charset        Charset meta tag(s).
+	 *     @type DOMElement[] $viewport       Viewport meta tag(s).
+	 *     @type DOMElement[] $amp_script_src <amp-script> source meta tags.
+	 *     @type DOMElement[] $other          Remaining meta tags.
+	 * }
+	 */
+	protected $meta_tags = [
+		self::TAG_CHARSET        => [],
+		self::TAG_VIEWPORT       => [],
+		self::TAG_AMP_SCRIPT_SRC => [],
+		self::TAG_OTHER          => [],
+	];
+
+	/**
+	 * Charset to use for AMP markup.
+	 *
+	 * @var string
+	 */
+	const AMP_CHARSET = 'utf-8';
+
+	/**
+	 * Viewport settings to use for AMP markup.
+	 *
+	 * @var string
+	 */
+	const AMP_VIEWPORT = 'width=device-width';
+
+	/**
+	 * Sanitize.
+	 */
+	public function sanitize() {
+		$this->xpath = new DOMXPath( $this->dom );
+		$this->head  = $this->ensure_head_is_present();
+		$charset     = $this->detect_charset();
+		$elements    = $this->dom->getElementsByTagName( static::$tag );
+
+		// Remove all nodes for easy reordering later on.
+		$elements = array_map(
+			static function ( $element ) {
+				return $element->parentNode->removeChild( $element );
+			},
+			iterator_to_array( $elements, false )
+		);
+
+		foreach ( $elements as $element ) {
+			/**
+			 * Meta tag to process.
+			 *
+			 * @var DOMElement $element
+			 */
+			if ( $element->hasAttribute( 'charset' ) ) {
+				$this->meta_tags[ self::TAG_CHARSET ][] = $element;
+			} elseif ( 'viewport' === $element->getAttribute( 'name' ) ) {
+				$this->meta_tags[ self::TAG_VIEWPORT ][] = $element;
+			} elseif ( 'amp-script-src' === $element->getAttribute( 'name' ) ) {
+				$this->meta_tags[ self::TAG_AMP_SCRIPT_SRC ][] = $element;
+			} else {
+				$this->meta_tags[ self::TAG_OTHER ][] = $element;
+			}
+		}
+
+		$this->ensure_charset_is_present( $charset );
+
+		if ( ! $this->is_correct_charset() ) { // phpcs:ignore Generic.CodeAnalysis.EmptyStatement
```

Comment:
I believe this will need to be done prior to the `DOMDocument` being constructed, but it didn't occur to me that perhaps `DOMDocument` allows for encoding to be changed after parsing. I don't think it facilitates this, however.

Reply:
DOMDocument works with UTF-8 internally, so I think we need to have it converted before sending it to DOMDocument, or otherwise we might already have messed it up via `loadHTML()`.

Reply:
Yes, I've put this into `AMP_DOM_Document` now.